kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -325,34 +362,41 @@ func (c *ipamController) handleNodeUpdate(kvp model.KVPair) {
 	if kvp.Value != nil {
 		n := kvp.Value.(*libapiv3.Node)
 		kn, err := getK8sNodeName(*n)
+
 		if err != nil {
-			log.WithError(err).Info("Unable to get corresponding k8s node name, skipping")
+			log.WithError(err).Info("Unable to get corresponding k8s node name")
+		}
 
-			// It's possible that a previous version of this node had an orchRef and so was added to the
-			// map. If so, we need to remove it.
-			if current, ok := c.kubernetesNodesByCalicoName[n.Name]; ok {
-				logrus.Warnf("Update mapping calico node -> k8s node. %s -> %s (previously %s)", n.Name, kn, current)
-				delete(c.kubernetesNodesByCalicoName, n.Name)
-			}
-		} else if kn != "" {
-			if current, ok := c.kubernetesNodesByCalicoName[n.Name]; !ok {
-				logrus.Debugf("Add mapping calico node -> k8s node. %s -> %s", n.Name, kn)
-				c.kubernetesNodesByCalicoName[n.Name] = kn
-			} else if current != kn {
-				logrus.Warnf("Update mapping calico node -> k8s node. %s -> %s (previously %s)", n.Name, kn, current)
-				c.kubernetesNodesByCalicoName[n.Name] = kn
-			}
-			// No change.
+		// Maintain mapping of Calico node to Kubernetes node. Ensure all Calico nodes have an entry in the map by
+		// assigning a value of "" for nodes that are not orchestrated by Kubernetes (1) or have a Kubernetes name of "" (2).
+		// Expects getK8sNodeName to return "", err for (1) and "", nil for (2).
+		if current, ok := c.kubernetesNodesByCalicoName[n.Name]; !ok {
+			log.Debugf("Add mapping calico node -> k8s node. %s -> %s", n.Name, kn)
+			c.kubernetesNodesByCalicoName[n.Name] = kn
+		} else if current != kn && !(err == nil && kn == "") {
```

Comment:
I'm not a fan of this delayed error handling - I think it makes this code a bit tricky to understand and isn't a great practice. Would rather make it a bit more verbose if that's what is needed. But, like I said in slack, I think we can get rid of this altogether if we treat "No orch ref" and "A k8s orch ref that has an empty node value" both to mean "Not a Kubernetes node"

Reply:
Agreed, I was on the fence about this too hence why I wanted to see if we could remove it.